doc/compiler/optimizer/IntroLoopOptimizations.md

```diff
@@ -0,0 +1,183 @@
+ Inlined code (fast path) (I)      call (slow path) (C)  Inlined code (fast path) (I')     call (slow path) (C')
+```
+For example, there is a loop with a non-loop invariant virtual guard and an invariant bound check inside the loop. After the loop is versioned with the bound check, there are two versions of the loop. One version is the fast/hot loop that does not have the bound check (`A`). Another one is the slow/cold loop that is a clone of the fast loop but with all the exception checks (`A’`). Both have the guards: `G` inside the fast loop, `G’` insides the slow loop.
+
+Loop Transfer takes the slow path (call path) out of the guard and redirects it to the slow loop. In the fast loop, if the virtual guard succeeds, it falls into the inlined code block (`I`). If it fails, it will not go into the call block (`C`) and it will go to `C’` in the slow loop.
+
+```
+                    Fast/Hot Loop (A)                                       Slow/Cold Loop (A')
+                               |                                                       |
+                               |                                                       |
+                               v                                                       v
+                            Guard (G)                                               Guard (G')
+                            /      |                                                /     \
+                           /       |                                               /       \
+                          v        |                                              v         v
+ Inlined code (fast path) (I)      |                     Inlined code (fast path) (I')     call (slow path) (C')
+                                   |                                                                ^
+                                   |                                                                |
+                                   +----------------------------------------------------------------+
+
+```
+The point of this optimization is to essentially remove the call from the fast loop by sending the slow path out of that guard in the fast/hot loop into the slow loop. After the transformation, the fast loop only has the fast path. It has the guard and its fast path, but if the guard fails, it goes into the slow loop and take the slow path (`C’`) in the slow loop. The slow path (`C`) is eliminated from the fast loop. When the call (`C`) is gone, it may open up opportunities for other optimizations in the fast loop.
+
+The bad thing about Loop Transfer is that it barges into the slow loop. It does not enter the slow loop in a well-behaved way. A loop should be entered from the entry block of the loop. That is the main property of a natural loop structure. A loop should not be entered at any place. It should be entered in a nice and structured way at the top of the loop.
```

Comment:
> that it barges

I think you meant to say "branches" here.

Reply:
Clarified on the above. I'll change it to branch since it is a more commonly used.

Reply:
Addressed in 641a90a